draft/0002-subdocapi.md

```diff
@@ -0,0 +1,351 @@
+# Meta
+
+ - RFC Name: SubDocument API for clients
+ - RFC ID: 0002
+ - Start Date: 2015-11-11
+ - Owner: Mark Nunberg
+ - Current Status: Draft
+
+# Summary
+
+This RFC proposes a client-side API for the Sub-Document operations which will
+be released in a future server version.
+
+# Motivation
+
+Expose the high performance subdocument API to developers
+
+# General Design
+
+## Terminology
+
+* Document: A document or item in the Memcached sense.
+* Path: A specific location in a JSON document. A path uses the path syntax
+  described in the server-side API.
+* Sub-Document: A section of a document, specified by the path.
+* Top-level: Document-level operations, for example `upsert`.
+
+## Overview
+
+The Sub-Document API exposes CRUD-like methods which operate on paths _within_
+documents. In these respects it is identical to CRUD methods except that it
+treats each path as its own "sub-document" (hence the name).
+
+Subdocument operations come in two flavors. One is the simple single-operation
+form which takes a single document ID and a single path. Another is a
+_multi-path_ form which retrieves or mutates _several_ paths within a document.
+
+Single-path methods are slightly more efficient and require less processing
+overhead than their multi-path equivalents (that is, even if you pass a single
+path to a multipath method, some additional processing overhead may be involved).
+
+Note that both 'multi-path' and 'single-path' APIs operate on the same
+document, and correspond to a single binary operation. This should not be
+confused with 'multi' operations in some SDKs which pipeline multiple binary
+requests in the network.
+
+## Objects
+
+MetaDocFragment
+
+This subclass contains the same fields as the MetaDoc, except the `value`
+may wish to be replaced by the `fragment` field. This is to avoid confusion
+regarding whether the value "truly" contains the entire document value, or
+merely a subdoc.
+
+The `MetaDoc` object itself is the return value from normal KV operations
+which contain the value, cas, mutation token, etc.
+
+This object is used as an example, some SDKs may wish to either use the
+'value' or 'document' field of the existing `MetaDoc` return value, or
+subclass it, replacing value with 'fragment'.
+
```

Comment:
One comment on naming: I was under the impression that in the SDK 2.0 initiative the doc is called `Document` and its content `content`? So for consistency reasons would make  `DocumentFragment` sense which has the same `content` field? So the user only needs to distinguish between a `Document` and a `DocumentFragment`.